src/libfsm/concat.c

```diff
@@ -67,10 +67,12 @@ fsm_concat(struct fsm *a, struct fsm *b,
 	fsm_setend(q, ea, 0);
 
 	/*
-	 * The canonical approach is to create epsilon transition(s) from the end
+	 * The textbook approach is to create epsilon transition(s) from the end
 	 * state of one FSM to the start state of the next FSM.
 	 *
-	 * TODO: diagram
+	 *     a: ⟶ ○ ┄ ◎
+	 *                     a b: ⟶ ○ ┄ ○ ⟶ ○ ┄ ◎
+	 *     b: ⟶ ○ ┄ ◎
```

Comment:
I like these a lot, overall.

But, whenever a state _symbol_ changes (e.g. to/from final) in the FSM composition, it makes it much harder to instantly see what state came from where. At least in those cases, it could help to include node labels.

Something roughly like:
```
	      a: ⟶ ○ ┄ ◎
                     a₁  a₂
	                      a b: ⟶ ○ ┄ ○ ⟶ ○ ┄ ◎
                                       a₁  a₂   b₁   b₂
	      b: ⟶ ○ ┄ ◎
                     b₁   b₂
```
More padding than in the above would probably be helpful, but you get the idea.

Reply:
I wondered about bracketing them perhaps. Like er...
```
a b: ⟶ ○ ┄ ○ ⟶ ○ ┄ ◎
      ╰──┬──╯ ╰──┬──╯
         a       b
```

Reply:
Ooh, I think that's an improvement, too.

My only concern is that the brackets might visually overpower the arrows, since they present longer (unbroken) lines than the arrow tails. But there's a similar concern with my sketch, where the state labels can be a bit noisy. Brackets would probably be ideal if there were a way to make them a slightly dimmed color.

What I do like about state labels is that my eye can bounce between two states and instantly see "a₂ was final before composition, but isn't any longer". In other words, the idea that a state in the new FSM is "the same" as one of the inputs is encoded in the preservation of the label, even if its properties (final or not) change (seen in the symbol).

That said, a bias is that I am personally _very_ used to such a convention from stuff like category theory, so it might just be clearer from familiarity. I guess it's worth looking at how these generalize to the other diagrams, too. Particularly `union`.

Reply:
Definitely replacing the `┄` with `○ ··· ◎`, that helps.